File: C++/Solution2.cpp
#include<iostream>
#include<vector>
#include<map>

using namespace std;

  struct ListNode {
      int val;
      ListNode *next;
      ListNode() : val(0), next(nullptr) {}
      ListNode(int x) : val(x), next(nullptr) {}
      ListNode(int x, ListNode *next) : val(x), next(next) {}
  };
 
class Solution2 {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* p = new ListNode();
        ListNode* head = p;
        int over = 0;
        while(l1 || l2){
            int ele1 = 0, ele2 = 0;
            if(l1 && l2){
                ele1 = l1 -> val;
                ele2 = l2 -> val;
                l1 = l1 -> next;
                l2 = l2 -> next;
            }
            else if(l1){
                ele1 = l1 -> val;
                l1 = l1 -> next;
            }
            else{
                ele2 = l2 -> val;
                l2 = l2 -> next;
            }
            int ele = ele1 + ele2 + over;
            p -> next = new ListNode(ele % 10);
            if(ele >= 10){
                over = 1;
            }
            else{
                over = 0;
            }
            p = p -> next;
        }
        if(over == 1){
            p -> next = new ListNode(1);
        }
        return head -> next;
    }
};

File: Java/other/Solution2.java
package other;
public class Solution2 {
    public static ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int[] sum1 = new int[100];
        int[] sum2 = new int[100];
        int i1 = 0;
        int i2 = 0;
        ListNode l1_copy = l1;
        ListNode l2_copy = l2;
        while(l1 != null){
            sum1[i1] = l1.val;
            i1 += 1;
            l1 = l1.next;
        }
        while(l2 != null){
            sum2[i2] = l2.val;
            i2 += 1;
            l2 = l2.next;
        }
        int over = 0;
        i1 -= 1;
        i2 -= 1;
        int j1 = 0;
        int j2 = 0;
        int sum = sum1[j1] + sum2[j2];
        ListNode result = new ListNode(sum % 10);
        ListNode copy = result;
        if(sum >= 10){
            over = 1;
        }
        j1 += 1;
        j2 += 1;
        while(j1 <= i1 || j2 <= i2){
            if(j1 > i1){
                copy.next = new ListNode((sum2[j2] + over) % 10);
                copy = copy.next;
                if(sum2[j2] + over >= 10){over = 1;}
                else{over = 0;}
                j2 += 1;
                continue;
            }
            else if(j2 > i2){
                 copy.next = new ListNode((sum1[j1] + over) % 10);
                copy = copy.next;
                if(sum1[j1] + over >= 10){over = 1;}
                else{over = 0;}
                j1 += 1;
                continue;
            }
            else{
                copy.next = new ListNode((sum1[j1] + sum2[j2] + over) % 10);
                copy = copy.next;
                if(sum1[j1] + sum2[j2] + over >= 10){over = 1;}
                else{over = 0;}
                j1 += 1;
                j2 += 1;
            }
        }
        if(over == 1){
            copy.next = new ListNode(1);
        }
        return result;
    }

    public static void main(String[] args){
        ListNode x1 = new ListNode(9);
        ListNode x2 = new ListNode(9);
        for(int i = 0; i <= 8; i++){
            x2 = new ListNode(9, x2);
        }
        x2 = new ListNode(1, x2);
        ListNode result = Solution2.addTwoNumbers(x1, x2);
        while(result != null){
            System.out.println(result.val);
            result = result.next;
        }
    }
}